Use ThrowHelper methods in more places

turbedi opened the pull request:
- Might help with https://github.com/ppy/osu-framework/issues/3470

ObjectDisposedException.ThrowIf(), ArgumentException.ThrowIfNullOrEmpty() and the ArgumentOutOfRangeException methods have been introduced in .NET 7/8

smoogipoo commented on osu.Framework/Bindables/BindableNumber.cs:
```diff
@@ -42,8 +42,7 @@ public T Precision
 
-                if (value.CompareTo(default) <= 0)
-                    throw new ArgumentOutOfRangeException(nameof(Precision), value, "Must be greater than 0.");
+                ArgumentOutOfRangeException.ThrowIfNegativeOrZero(value.CompareTo(default));
```
Can this be `ThrowIfNegativeOrZero(value)`?

turbedi replied:
Yes should be fine. Will change it.

Susko3 replied:
Having this done via `CompareTo()` results in the actual value being misrepresented. Eg. setting a `BindableFloat.Precision` to `-5f` will show `-1` as the actual value.

```cs
System.ArgumentOutOfRangeException : value.CompareTo(default) ('-1') must be a non-negative and non-zero value. (Parameter 'value.CompareTo(default)')
Actual value was -1.
```

This is better handled via generic math (could just call `ArgumentOutOfRangeException.ThrowIfNegativeOrZero(value)`). But that is a breaking change and should not be done in this PR, so best to revert for now.

turbedi replied:
Done

Susko3 commented on osu.Framework/Graphics/Containers/GridContainer.cs:
```diff
@@ -420,11 +420,8 @@ public class Dimension
-            if (minSize > maxSize)
-                throw new ArgumentOutOfRangeException(nameof(minSize), $"Must be less than {nameof(maxSize)}.");
+            ArgumentOutOfRangeException.ThrowIfNegative(minSize);
+            ArgumentOutOfRangeException.ThrowIfGreaterThan(minSize, maxSize);
```
Just noting that `maxSize` no will longer appear in the error message.

master:
```cs
System.ArgumentOutOfRangeException : Must be less than maxSize. (Parameter 'minSize')
```

This PR:
```cs
System.ArgumentOutOfRangeException : minSize ('20') must be less than or equal to '10'. (Parameter 'minSize')
Actual value was 20.
```

Susko3 commented on osu.Framework/Testing/Drawables/TestGroupButton.cs:
```diff
@@ -45,8 +45,7 @@ public TestGroupButton(Action<Type> loadTest, TestGroup group)
 
-            if (tests.Length == 0)
-                throw new ArgumentOutOfRangeException(nameof(group), tests.Length, "Type array must not be empty!");
+            ArgumentOutOfRangeException.ThrowIfZero(tests.Length);
```
Should be reverted, `tests.Length` is not an argument of this method. The error message doesn't make much sense.
```cs
0 ('tests.Length') must be a non-zero value.
```

Susko3 commented on osu.Framework/Utils/IncrementalBSplineBuilder.cs:
```diff
@@ -140,8 +138,7 @@ public float CornerThreshold
             {
-                if (cornerThreshold < 0)
-                    throw new ArgumentOutOfRangeException(nameof(value), "CornerThreshold must not be negative.");
+                ArgumentOutOfRangeException.ThrowIfNegative(cornerThreshold);
```
Wait this is actually wrong on master, it's checking the old `cornerThreshold` and not the newly-set `value`.
Should be changed to

```cs
ArgumentOutOfRangeException.ThrowIfNegative(value);
```

Susko3 commented on osu.Framework/Utils/RNG.cs:
```diff
@@ -61,8 +60,7 @@ public static double NextDouble(double maxValue)
         {
-            if (minValue > maxValue)
-                throw new ArgumentOutOfRangeException(nameof(minValue), "The given minimum value must be less than or equal to the given maximum value.");
+            ArgumentOutOfRangeException.ThrowIfGreaterThan(minValue, maxValue);
```
See other comment about `maxValue`/`maxSize` no longer appearing in the error message. Same applies for the other usage of `ThrowIfGreaterThan()` in this file.

Susko3 commented on osu.Framework/Utils/IncrementalBSplineBuilder.cs:
```diff
@@ -117,8 +116,7 @@ public float Tolerance
             {
-                if (tolerance < 0)
-                    throw new ArgumentOutOfRangeException(nameof(value), "Tolerance must not be negative.");
+                ArgumentOutOfRangeException.ThrowIfNegative(tolerance);
```
Should be:

```cs
ArgumentOutOfRangeException.ThrowIfNegative(value);
```

Susko3 reviewed (changes requested):
Seems ok but some error messages have changed for the worse.

turbedi commented:
I've addressed some of the comments. Should I also revert all usages of ThrowIfGreaterThan() to retain the max value in the error message, or is it fine like that?

smoogipoo commented:
The other path I can suggest is to make an `ExceptionUtils.ValidateRange()`. Otherwise, I'm _weakly_ okay with the change in messaging but I'd want other reviews.

frenzibyte commented:
I personally don't feel the exception message matters that much. Sure it might be a little bit cryptic, but the "must be greater than"/"must be lower than" is already giving it out I think.

bdach commented:
My view on the `ThrowIfGreaterThan()` thing is that I'd be fine with the max value not being printed *only* if it's a constant. So in cases where the max value isn't statically known just by looking at the direct call site it should print the max in the exception message.

In favour of revert.

turbedi commented:
All comments have been addressed.
